File: Develop/frontend/src/services/systemCodeService.ts
/**
 * 系統代碼服務
 *
 * 注意: axios 攔截器會自動從 localStorage 讀取 txn_token 並添加到所有請求的 header
 * Service 層不需要手動傳遞 txn_token 參數
 */

import axios from '../api/axios';
import { I18nField } from '../types';

const BASE_URL = '/api/system_codes';

export interface SystemCode {
  id: number;
  code_type: string;
  code_type_name: I18nField;
  code: string;
  code_name: I18nField;
  order: number;
  is_active: boolean;
  note1?: string;
  note2?: string;
  note3?: string;
  note4?: string;
  note5?: string;
  edit_by: number;
  created_at: string;
  updated_at: string;
}

export interface SystemCodeCreate {
  code_type: string;
  code_type_name: I18nField;
  code: string;
  code_name: I18nField;
  order?: number;
  is_active?: boolean;
  note1?: string;
  note2?: string;
  note3?: string;
  note4?: string;
  note5?: string;
}

export interface SystemCodeUpdate {
  code_type?: string;
  code_type_name?: I18nField;
  code?: string;
  code_name?: I18nField;
  order?: number;
  is_active?: boolean;
  note1?: string;
  note2?: string;
  note3?: string;
  note4?: string;
  note5?: string;
}

/**
 * 取得系統代碼列表
 * axios 攔截器會自動添加 X-Txn-Token header
 */
export const getSystemCodes = async (params?: {
  code_type?: string;
  code?: string;
  is_active?: boolean;
  search?: string;
}): Promise<SystemCode[]> => {
  const response = await axios.get<SystemCode[]>(`${BASE_URL}/`, { params });
  return response.data;
};

/**
 * 取得系統代碼資訊
 * axios 攔截器會自動添加 X-Txn-Token header
 */
export const getSystemCode = async (codeId: number): Promise<SystemCode> => {
  const response = await axios.get<SystemCode>(`${BASE_URL}/${codeId}`);
  return response.data;
};

/**
 * 建立系統代碼
 * axios 攔截器會自動添加 X-Txn-Token header
 */
export const createSystemCode = async (data: SystemCodeCreate): Promise<SystemCode> => {
  const response = await axios.post<SystemCode>(`${BASE_URL}/`, data);
  return response.data;
};

/**
 * 更新系統代碼
 * axios 攔截器會自動添加 X-Txn-Token header
 */
export const updateSystemCode = async (
  codeId: number,
  data: SystemCodeUpdate
): Promise<SystemCode> => {
  const response = await axios.put<SystemCode>(`${BASE_URL}/${codeId}`, data);
  return response.data;
};

/**
 * 刪除系統代碼
 * axios 攔截器會自動添加 X-Txn-Token header
 */
export const deleteSystemCode = async (codeId: number): Promise<void> => {
  await axios.delete(`${BASE_URL}/${codeId}`);
};

/**
 * 根據代碼類別查詢
 * axios 攔截器會自動添加 X-Txn-Token header
 */
export const getSystemCodesByType = async (
  codeType: string,
  activeOnly: boolean = true
): Promise<SystemCode[]> => {
  const params = new URLSearchParams({ active_only: String(activeOnly) });
  const response = await axios.get<SystemCode[]>(`${BASE_URL}/type/${codeType}?${params.toString()}`);
  return response.data;
};

File: Develop/backend/app/services/systemcode_service.py
"""
SystemCode Service
系統代碼服務層
"""

import logging
from sqlalchemy.orm import Session
from sqlalchemy import or_, text, cast, String
from typing import List, Optional

from app.models.systemcode import SystemCode
from app.schemas.systemcode import SystemCodeCreate, SystemCodeUpdate, SystemCodeQuery

logger = logging.getLogger(__name__)


class SystemCodeService:
    """系統代碼服務"""

    @staticmethod
    def get_all(db: Session, query: Optional[SystemCodeQuery] = None) -> List[SystemCode]:
        """查詢系統代碼列表"""
        q = db.query(SystemCode)

        if query:
            if query.search:
                # 綜合搜尋：code_type、code、所有語系名稱、note1~5
                search_term = f"%{query.search}%"
                search_filter = or_(
                    SystemCode.code_type.ilike(search_term),
                    SystemCode.code.ilike(search_term),
                    # JSONB 多語系欄位：搜尋所有語系的值
                    cast(SystemCode.code_type_name, String).ilike(search_term),
                    cast(SystemCode.code_name, String).ilike(search_term),
                    SystemCode.note1.ilike(search_term),
                    SystemCode.note2.ilike(search_term),
                    SystemCode.note3.ilike(search_term),
                    SystemCode.note4.ilike(search_term),
                    SystemCode.note5.ilike(search_term),
                )
                q = q.filter(search_filter)
            else:
                if query.code_type:
                    q = q.filter(SystemCode.code_type.ilike(f"%{query.code_type}%"))
                if query.code:
                    q = q.filter(SystemCode.code.ilike(f"%{query.code}%"))

            if query.is_active is not None:
                q = q.filter(SystemCode.is_active == query.is_active)

        return q.order_by(
            SystemCode.code_type,
            SystemCode.order,
            SystemCode.code
        ).all()

    @staticmethod
    def get_by_id(db: Session, code_id: int) -> Optional[SystemCode]:
        """根據ID查詢系統代碼"""
        return db.query(SystemCode).filter(SystemCode.id == code_id).first()

    @staticmethod
    def create(db: Session, code_data: SystemCodeCreate, user_id: int) -> SystemCode:
        """建立系統代碼"""
        db_code = SystemCode(
            **code_data.model_dump(),
            edit_by=user_id
        )
        db.add(db_code)
        db.commit()
        db.refresh(db_code)
        logger.info(f"[SystemCodeService] Created system code: {db_code.id} by user {user_id}")
        return db_code

    @staticmethod
    def update(db: Session, code_id: int, code_data: SystemCodeUpdate, user_id: int) -> Optional[SystemCode]:
        """更新系統代碼"""
        db_code = SystemCodeService.get_by_id(db, code_id)
        if not db_code:
            return None

        update_data = code_data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_code, field, value)

        db_code.edit_by = user_id
        db.commit()
        db.refresh(db_code)
        logger.info(f"[SystemCodeService] Updated system code: {code_id} by user {user_id}")
        return db_code

    @staticmethod
    def delete(db: Session, code_id: int) -> bool:
        """刪除系統代碼"""
        db_code = SystemCodeService.get_by_id(db, code_id)
        if not db_code:
            return False

        db.delete(db_code)
        db.commit()
        logger.info(f"[SystemCodeService] Deleted system code: {code_id}")
        return True

    @staticmethod
    def get_by_type(db: Session, code_type: str, active_only: bool = True) -> List[SystemCode]:
        """根據代碼類別查詢系統代碼"""
        q = db.query(SystemCode).filter(SystemCode.code_type == code_type)

        if active_only:
            q = q.filter(SystemCode.is_active == True)

        return q.order_by(SystemCode.order, SystemCode.code).all()
